application-licensing-licensor/application-licensing-licensor-api/src/main/java/com/xwiki/licensing/internal/upgrades/notifications/ExtensionAutoUpgradedEventDisplayer.java

```diff
@@ -56,44 +53,45 @@
 @Named(ExtensionAutoUpgradedEventDisplayer.NAME)
 public class ExtensionAutoUpgradedEventDisplayer implements NotificationDisplayer
 {
-    protected static final String CODE_SPACE = "Code";
-
     protected static final String NAME = "ExtensionAutoUpgradedEventDisplayer";
 
     protected static final List<String> EVENTS = Arrays.asList(ExtensionAutoUpgradedEvent.class.getCanonicalName(),
         ExtensionAutoUpgradedFailedEvent.class.getCanonicalName());
 
-    protected static final LocalDocumentReference DISPLAYER_DOC =
-        new LocalDocumentReference(Arrays.asList("Licenses", CODE_SPACE), NAME);
+    protected static final LocalDocumentReference LICENSOR_DOC = new LocalDocumentReference("Licenses", "WebHome");
 
-    protected static final LocalDocumentReference DISPLAYER_OBJ =
-        new LocalDocumentReference(Arrays.asList("XWiki", "Notifications", CODE_SPACE), "NotificationDisplayerClass");
+    protected static final String EVENT_BINDING_NAME = "event";
 
     @Inject
     private TemplateManager templateManager;
 
     @Inject
     private Provider<XWikiContext> contextProvider;
 
+    @Inject
+    private ScriptContextManager scriptContextManager;
+
     @Inject
     private Logger logger;
 
     @Override
     public Block renderNotification(CompositeEvent eventNotification) throws NotificationException
     {
         XWikiContext xcontext = contextProvider.get();
+        ScriptContext scriptContext = this.scriptContextManager.getScriptContext();
+        Template customTemplate = this.templateManager.getTemplate("extensionAutoUpgraded/extensionAutoUpgraded.vm");
 
         try {
-            XWikiDocument displayerDoc = xcontext.getWiki().getDocument(DISPLAYER_DOC, xcontext);
-            BaseObject displayerObj = displayerDoc.getXObject(DISPLAYER_OBJ, 0);
+            // Set a document in the context to act as the current document when the template is rendered.
+            xcontext.setDoc(xcontext.getWiki().getDocument(LICENSOR_DOC, xcontext));
+
+            // Bind the event to some variable in the velocity context.
+            scriptContext.setAttribute(EVENT_BINDING_NAME, eventNotification, ScriptContext.ENGINE_SCOPE);
 
-            Template customTemplate = templateManager.createStringTemplate(
-                displayerObj.getStringValue("notificationTemplate"), displayerDoc.getAuthorReference());
             return (customTemplate != null) ? this.templateManager.execute(customTemplate)
                 : this.templateManager.execute("notification/default.vm");
```

Comment:
The fallback on the default template means that a missing custom template is somehow valid (no a problem). But we bundle the custom template in the same JAR as this code so it should always be found, unless there's a serious problem, in which case we should be aware of. I think the fallback make sense when this code doesn't "own" the custom template, i.e. when the custom template is provided by some other module and it _may_ be missing. And since the custom template is own by this code I wouldn't do the fallback. Unless this is a best practice for notification displayers that I'm not aware of.

Reply:
To be clear, the fallback doesn't hurt and can be considered as bulletproofing. My issue is that it sends the wrong message that the custom template _may_ be missing, while it should _never_ be missing since it is bundled with the code.